### ISSUE TITLE
Add subsection to Promises to explain incumbent settings object tracking in callbacks 

### ISSUE THREAD

**Issue by syg** [mdn/sprints#3495]
## URL(s)
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise

## Request type

- [x] New documentation
- [ ] Correction or update

## Details

We are working on describing the [incumbent settings object](https://html.spec.whatwg.org/#incumbent) tracking behavior in Promise callbacks in both the JavaScript standard and the HTML standard. Currently, Firefox is the only implementation that tracks the incumbent. This warrants a note in the HTML standard that currently only a single implementation supports the feature. The HTML standard records compat information via MDN links in the sidebar, and for that, we need an MDN article describing the incumbent behavior.

The same behavior also extends to callbacks passed to `FinalizationRegistry` as part of the new WeakRef JS feature.

cc @codehag who's helping out from the Firefox side.

**Comment by syg** [comment]
Pinging this.

Now that WeakRefs and FinalizationRegistry are stage 4 and been merged into [ecma262](https://tc39.es/ecma262/#sec-finalization-registry-constructor), it'd be good to also merge the HTML integration PR. That PR is currently blocked on MDN having a linkable article for incumbent handling in Promises and FinalizationRegistry callbacks for the compat warning that Firefox is the only one to implement the behavior.

**Comment by Elchi3** [comment]
Thanks for your issue!

I don't understand incumbent from reading the spec. If you have any explaining material other than the spec, it would be useful to learn more. If such material doesn't exist, then 
- adding to the MDN promise article sounds great.
- or there is also https://developer.mozilla.org/en-US/docs/Web/JavaScript/EventLoop which might be relevant here,
- or we can create an entire new page dedicated to incumbent and I can give you page creation permission on MDN.

>  HTML standard records compat information via MDN 

Right, this is done via https://github.com/mdn/browser-compat-data and we could, for example, add a new sub feature to https://github.com/mdn/browser-compat-data/blob/master/javascript/builtins/Promise.json describing incumbent. 

> compat warning that Firefox is the only one to implement the behavior.

Do you or @codehag have a Firefox bug handy for this? When did this ship? Does the bug have "dev-doc-needed" set?

**Comment by syg** [comment]
I unfortunately do not have any explanatory material better than the spec text I linked. It is unfortunately a very arcane concept. I think the 3 examples in [here](https://html.spec.whatwg.org/#incumbent:incumbent-settings-object-6:~:text=The%20following%20series%20of%20examples%20is,the%20definition%20of%20the%20incumbent%20concept%3A) is the best we got.

I think to start, just adding the Promise article sounds good.

As for when did Firefox ship this behavior, I'll defer @codehag to answer. bz has historically been the keeper of incumbent semantics, and he's off to other ventures. :)

**Comment by codehag** [comment]
Hi, sorry had this on my todo list yesterday and didn't get to it. 

We introduced incumbent realms with our promise implementation. I don't know how specific our documentation about when this shipped needs to be, but it was introduced in this bug: https://bugzilla.mozilla.org/show_bug.cgi?id=911216 and this patch specifically: https://bug911216.bmoattachments.org/attachment.cgi?id=8761429

This _is_ browser specific behavior. It isn't something that will be present in other JS implementations. I can write up an explanatory text if it would be helpful.

**Comment by chrisdavidmills** [comment]
@codehag that would be super helpful, thanks Yulia!

**Comment by codehag** [comment]
Here is the text for incumbent realms on promises: https://wiki.developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise#Incumbent_Realm

**Comment by sideshowbarker** [comment]
I now wonder if in MDN we should instead be titling this **Tracking the incumbent settings object** (or if we want to be more accurate, **Tracking the realm of the incumbent settings object**).

Or alternatively, the https://github.com/whatwg/html/pull/5722/files HTML spec patch could be rewritten to normatively use the term *incumbent realm* (as shorthand for “the realm of the incumbent settings object”).

I say all that because I now notice that nowhere in the https://github.com/whatwg/html/pull/5722/files HTML spec patch for this feature as currently written does the patch actually use the term *incumbent realm*. (In fact the patch doesn’t even ever refer to the “the realm of the incumbent settings object” — but instead just talks about saving and restoring the incumbent settings object itself (not the _realm_ of the incumbent settings object).

And while it’s true that at https://html.spec.whatwg.org/multipage/webappapis.html#concept-incumbent-realm, the current HTML spec does define the term *incumbent realm*, the spec itself never even normatively references that term — it doesn’t actually use that term in any algorithms or other requirements (instead only in a non-normative example).

So anyway, without some further change for this somewhere, it seems like we have potentially confusing mismatch between the https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise article’s terminology, and the terminology that’s used in the spec patch.

**Comment by codehag** [comment]
Ok, I have updated the naming everywhere and added a description of the incumbent settings object to the mdn documentation.

**Comment by sideshowbarker** [comment]
> Ok, I have updated the naming everywhere and added a description of the incumbent settings object to the mdn documentation.

Thanks, the update looks perfect

> IIUC Its more accurate, part of the issue is with the settings object (ie, tracking the current url)

Yeah I think the updated description (along with the example code) in the MDN article does an exceptional job of explaining it clearly in a way that makes it relevant to web-developer needs

**Comment by Elchi3** [comment]
Looks like this got done 🎉 https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise#Incumbent_settings_object_tracking

**Comment by danieltroger** [comment]
Yo I don't understand the stuff in MDN.

First of all it's confusing to me whether this actually contributes to the example:
`Promise.resolve(undefined).then(bound);`
(.then queues bound to the next iteration of the microtask queue)

Or if `(0,bound)()` would work just as well? (Executes it instantly)

Also in the example of x.html it says `document.querySelector("#text").textContent = "hello";` but there is no element with the id `text` in any of the snippets.

And then idk I don't understand the idea. What thing has to be tracked? `frames[0]` is bound to the postMessage function and that's basically a new function, no? Why does something have to be tracked? What's the thing that's tracked?

**Comment by scottyinthematrix** [comment]
I don't understand this part either, to me it's quite abrupt that such a thing showing up under the topic of Promise, anyway, what's the relationship between the two, should I understand it before understanding the Promise? Maybe a separate topic for that is better?
